zjit/src/hir.rs

```diff
@@ -138,6 +138,40 @@ impl Invariant {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum SpecialObjectType {
+    VMCore = 1,
+    CBase = 2,
+    ConstBase = 3,
+}
+
+impl From<u32> for SpecialObjectType {
+    fn from(value: u32) -> Self {
+        match value {
+            1 => SpecialObjectType::VMCore,
+            2 => SpecialObjectType::CBase,
+            3 => SpecialObjectType::ConstBase,
```

Comment:
I think you can use `VM_SPECIAL_OBJECT_*` constants in cruby_bindings.inc instead of magic numbers 1, 2, and 3.